sample-prep/Biomek_pooling_VI.py

```diff
@@ -1,22 +1,36 @@
 '''
-Script to create csv-file for pooling on Biomek pipetting robot (post PCR lab)
-All calculations for pooling have been done with a LLTK evaluateDynamicExpression script, so we can extract ready-to-use UDFs from the process/step.
-Future updates: Also include LLTK calculations in this script?
+Script to create csv-file for pooling on Biomek pipetting robot and a pooling placement map as pdf for reference and/or manual pooling
+
+Species, Genome Size, Project Account are copied to the derived input using an LLTK script at step entry. Pooling volume calculations done by this script.
+At step entry, another script will check that input is max 2 plates. And no tubes allowed (at the moment) - change if needed in future. Will need major rewrite.
+
+author: Magnus Leithaug
 '''
 
 from genologics.lims import *
 from genologics import config
 import sys
 import re
+from plotnine import ggplot, geom_point, aes, geom_text, scale_y_discrete, scale_x_discrete, theme_bw, ggtitle, xlab, ylab, theme, scale_color_manual, save_as_pdf_pages, scale_color_identity
+import pandas as pd
+
+#sort_key function (for sorting the input lists)
+
+def sort_key(elem):
+    container_name = elem.container.name
+    container, well = elem.location
+    row, col = well.split(":")
+    return (container_name, int(col), row)
+
 
-def main(process_id, file_id):
+def main(process_id, file_id, pdf_file_id):
     
-    #load lims instance and process
+    #LOAD LIMS INSTANCE AND PROCESS
 
     lims = Lims(config.BASEURI, config.USERNAME, config.PASSWORD)
     process = Process(lims, id=process_id)
 
```

Comment:
What isi process here? Is this the db connection? If so, is the normal name in epps code for this?

Reply:
Here, the variable 'process' will store the class for the process, which is the APIs name for a step (type:  <class 'genologics.entities.Process'> ). Tapping into the class we can most notably access input-output maps (process.input_output_maps), with output and input uri`s, which in turn contain the udf data. From process we can also get udf`s that is related to the step itself. This could for example be a toggle switch you have to select before running the script (accessed with process.udf.get('UDF NAME')).